compiler/rustc_codegen_ssa/src/back/archive.rs

```diff
@@ -54,6 +57,38 @@ pub trait ArchiveBuilderBuilder {
         dll_imports: &[DllImport],
         tmpdir: &Path,
     ) -> PathBuf;
+
+    fn unpack_archive(
+        &self,
+        archive: &Path,
+        outdir: &Path,
+        staticlib_suffix: &str,
+    ) -> Result<(), String> {
+        let message = |msg: &str, e| format!("{} '{}': {}", msg, &archive.display(), e);
+        let archive_map = unsafe {
+            Mmap::map(
+                File::open(archive).map_err(|e| message("failed to open file", e.to_string()))?,
+            )
+            .map_err(|e| message("failed to mmap file", e.to_string()))?
+        };
+        let archive = ArchiveFile::parse(&*archive_map)
+            .map_err(|e| message("failed to parse archive", e.to_string()))?;
+
+        let data: &[u8] = archive_map.as_ref();
+        for entry in archive.members() {
+            let entry = entry.map_err(|e| message("failed to read entry", e.to_string()))?;
+            let data = entry.data(data).unwrap();
+            let name = String::from_utf8(entry.name().to_vec())
+                .unwrap_or_else(|e| message("failed to convert name", e.to_string()));
+            if !name.to_string().ends_with(staticlib_suffix) {
+                continue; // We need to unpack only native libraries.
```

Comment:
This won't work with bundled `+verbatim` libraries though, unless they also have the `staticlib_suffix` extension.
At least `verbatim` is not stable yet.
Could you add a FIXME comment mentioning this issue?